Changes on the linear-elasticity problem simulation using PSYDAC.

ArasuCandassamy opened the pull request:
This pull request addresses the linear elasticity problem and its simulation using Psydac. The initial Jupyter notebook was not functional. This PR introduces two new notebooks, each implementing a different formulation.

The first notebook employs the pure displacement formulation to solve the PDE with one vector field (the displacement) as the unknown. This approach is similar to the previous notebook but has been corrected to ensure the Python script works properly.

The second notebook utilizes a mixed formulation known as the displacement-pressure formulation. This method involves one vector field (the displacement) and one scalar field (the pressure) as unknowns.

ArasuCandassamy committed:
Commit changes on the linear-elasticity problem simulation using PSYDAC.

ArasuCandassamy committed:
clean up linear-elasticity-v2.ipynb.

ArasuCandassamy committed:
Correct comments in the code

yguclu committed:
Merge branch 'master' into linear-elasticity-fixes

yguclu commented on chapter2/linear-elasticity.ipynb:
I guess that a markdown file could be used here, instead of an IPython Notebook with a single markdown cell. The file would be called `linear-elasticity-intro.md`. See `intro.md` for an example. Please let me know if that works.

Further, your section "The PDE problem" is currently repeated in both the v1 and v2 notebooks. I like that section, but I think that it should not be repeated. Instead I suggest to move it to this markdown file, and remove it from the two notebooks. You can use a link to refer to this section from the notebooks.

ArasuCandassamy replied:
The issue occuring after the cell 
```
equation_h.set_solver('gmres', info=False, tol=1e-8)
uh = equation_h.solve(mu=mu_val, kappa=kappa_val)
```
was due to an excessively long runtime. This has been resolved by reducing the number of cells from 8 to 6.

Additionally, the section "The PDE problem" has been moved to a Markdown file and removed from both notebooks.

yguclu replied:
All right. Just a comment about how PR reviewing works...

It is good practice to let the person who started a conversation mark it as "resolved", especially if this person is reviewing the PR. It also helps if you tell the reviewer which commit has addressed their comment.

yguclu reviewed (changes requested):
There is a problem in both notebooks. After this cell
```python
equation_h.set_solver('gmres', info=False, tol=1e-8)
uh = equation_h.solve(mu=mu_val, kappa=kappa_val)
```
we get
```python
---------------------------------------------------------------------------
KeyboardInterrupt                         Traceback (most recent call last)
Cell In[4], line 2
      1 equation_h.set_solver('gmres', info=False, tol=1e-8)
----> 2 uh = equation_h.solve(mu=mu_val, kappa=kappa_val)

File /Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/psydac/api/equation.py:198, in DiscreteEquation.solve(self, **kwargs)
    196 def solve(self, **kwargs):
--> 198     self.assemble(**kwargs)
    200     # Free arguments of current equation
    201     free_args = set(self.lhs.free_args + self.rhs.free_args)

File /Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/psydac/api/equation.py:178, in DiscreteEquation.assemble(self, **kwargs)
    176 # Matrix (left-hand side)
    177 if assemble_lhs:
--> 178     A = self.lhs.assemble(reset=True, **kwargs)
    179     if self.bc:
    180         apply_essential_bc(A, *self.bc)

File /Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/psydac/api/fem.py:533, in DiscreteBilinearForm.assemble(self, reset, **kwargs)
    530 if reset:
    531     reset_arrays(*self.global_matrices)
--> 533 self._func(*args, *self._threads_args)
    534 if self._matrix and self._update_ghost_regions:
    535     self._matrix.exchange_assembly_data()

File ~/work/IGA-Python/IGA-Python/chapter2/__psydac__/dependencies_n2tehl4j.py:123, in assemble_matrix_n2tehl4j(global_test_basis_v_1, global_test_basis_v_2, global_test_basis_v_3, global_trial_basis_u_1, global_trial_basis_u_2, global_trial_basis_u_3, global_span_v_1, global_span_v_2, global_span_v_3, global_x1, global_x2, global_x3, test_v_p1, test_v_p2, test_v_p3, trial_u_p1, trial_u_p2, trial_u_p3, n_element_1, n_element_2, n_element_3, k1, k2, k3, pad1, pad2, pad3, g_mat_u_0_v_0_n2tehl4j, g_mat_u_1_v_0_n2tehl4j, g_mat_u_2_v_0_n2tehl4j, g_mat_u_0_v_1_n2tehl4j, g_mat_u_1_v_1_n2tehl4j, g_mat_u_2_v_1_n2tehl4j, g_mat_u_0_v_2_n2tehl4j, g_mat_u_1_v_2_n2tehl4j, g_mat_u_2_v_2_n2tehl4j, mu, kappa)
    121             contribution_v_2_u_0_n2tehl4j += temp_v_0_u_0_2*u_0_x3*v_2_x1 + temp_v_0_u_0_6*u_0_x1
    122             contribution_v_2_u_1_n2tehl4j += temp_v_0_u_0_2*u_1_x3*v_2_x2 + temp_v_0_u_0_6*u_1_x2
--> 123             contribution_v_2_u_2_n2tehl4j += temp_v_0_u_0_1*temp_v_0_u_0_7 + temp_v_0_u_0_2*temp_v_0_u_0_7 + temp_v_0_u_0_2*(temp_v_0_u_0_7 + u_2_x1*v_2_x1 + u_2_x2*v_2_x2)
    127 l_mat_u_0_v_0_n2tehl4j[i_basis_1,i_basis_2,i_basis_3,2 - i_basis_1 + j_basis_1,2 - i_basis_2 + j_basis_2,2 - i_basis_3 + j_basis_3] = contribution_v_0_u_0_n2tehl4j
    128 l_mat_u_1_v_0_n2tehl4j[i_basis_1,i_basis_2,i_basis_3,2 - i_basis_1 + j_basis_1,2 - i_basis_2 + j_basis_2,2 - i_basis_3 + j_basis_3] = contribution_v_0_u_1_n2tehl4j

KeyboardInterrupt: 
```

ArasuCandassamy committed:
Bug correction on notebooks. Replace the introduction page by a Markdown file and remove them from notebooks

ArasuCandassamy committed:
Merge remote-tracking branch 'origin/linear-elasticity-fixes' into linear-elasticity-fixes

yguclu committed:
Fix inline math in linear-elasticity-intro.md

Make sure that all inline math can be displayed correctly in the GitHub preview for markdown files.

ArasuCandassamy committed:
Provide a more general description of the PDE and clarify notations across notebooks.

ArasuCandassamy committed:
Cleaning notebooks

yguclu commented on chapter2/linear-elasticity-intro.md:
```diff
@@ -0,0 +1,157 @@
+    \sigma(u) &= C : \epsilon(u) & \text{in } \Omega \\
+\end{align}
+
+$$
```
This does not show properly on GitHub, which makes it hard to review the PR. Given that the `$$` delimiters enter math mode, I suppose that one should use `aligned` instead of `align`. Probably we should avoid empty lines, too.
```suggestion
$$
\begin{aligned}
    -\nabla \cdot \sigma(u) &= f & \text{in } \Omega \\
    \sigma(u) &= C : \epsilon(u) & \text{in } \Omega \\
\end{aligned}
$$
```

yguclu commented on chapter2/linear-elasticity-intro.md:
```diff
@@ -0,0 +1,157 @@
+    \sigma(u) \cdot n &= g_T & \text{on } & \partial \Omega_T
+\end{align}
+
+$$
```
See previous comment.
```suggestion
$$
\begin{aligned}
    -\nabla \cdot \sigma (u) &= f & \text{in } & \Omega \\
    u &= 0 & \text{on } & \partial \Omega_D \\
    \sigma(u) \cdot n &= g_T & \text{on } & \partial \Omega_T
\end{aligned}
$$
```

yguclu commented on chapter2/linear-elasticity-intro.md:
```diff
@@ -0,0 +1,157 @@
+
+\int_{\Omega} - \nabla \cdot \sigma(u) \cdot v \, \mathrm{d} x = \int_{\Omega} f \cdot v \, \mathrm{d} x
+
+$$
```
Here the problem must be the empty lines...
```suggestion
$$
\int_{\Omega} - \nabla \cdot \sigma(u) \cdot v \, \mathrm{d} x = \int_{\Omega} f \cdot v \, \mathrm{d} x
$$
```

yguclu commented on chapter2/linear-elasticity-intro.md:
```diff
@@ -0,0 +1,157 @@
+
+Integrating the term $ \nabla \cdot \sigma(u) \cdot v $ by parts, considering boundary conditions, we obtain:
+
+$$
+
+\int_{\Omega} \sigma(u) : \nabla v \, \mathrm{d} x = \int_{\Omega} f \cdot v \, \mathrm{d} x + \int_{\partial \Omega_T} g_T \cdot v \, \mathrm{d} s
+
+$$
```
```suggestion
$$
\int_{\Omega} \sigma(u) : \nabla v \, \mathrm{d} x = \int_{\Omega} f \cdot v \, \mathrm{d} x + \int_{\partial \Omega_T} g_T \cdot v \, \mathrm{d} s
$$
```

yguclu commented on chapter2/linear-elasticity-intro.md:
```diff
@@ -0,0 +1,157 @@
+
+By using the symmetry of the stress tensor $ \sigma $ and its definition from $(2)$, we can notice that :
+
+$$
+
+\begin{aligned}
+\int_{\Omega} \sigma(u) : \nabla v \, \mathrm{d} x &= \int_{\Omega} \sigma(u) : \epsilon(v) \, \mathrm{d} x = \int_{\Omega} C : \epsilon(u) : \epsilon(v) \, \mathrm{d} x \\ &= \int_{\Omega} \epsilon(u) : C : \epsilon(v) \, \mathrm{d} x
+\end{aligned}
+
+$$
```
```suggestion
$$
\begin{aligned}
\int_{\Omega} \sigma(u) : \nabla v \, \mathrm{d} x &= \int_{\Omega} \sigma(u) : \epsilon(v) \, \mathrm{d} x = \int_{\Omega} C : \epsilon(u) : \epsilon(v) \, \mathrm{d} x \\ &= \int_{\Omega} \epsilon(u) : C : \epsilon(v) \, \mathrm{d} x
\end{aligned}
$$
```

yguclu commented on chapter2/linear-elasticity-intro.md:
```diff
@@ -0,0 +1,157 @@
+
+This leads to the following variational formulation:
+
+$$
+
+\boxed{
+\begin{aligned}
+&\text{Find } u \in V \text{ such that:} \\
+&\qquad a(u, v) = L(v) \quad \forall v \in V
+\end{aligned}
+}
+
+$$
```
```suggestion
$$
\boxed{
\begin{aligned}
&\text{Find } u \in V \text{ such that:} \\
&\qquad a(u, v) = L(v) \quad \forall v \in V
\end{aligned}
}
$$
```

yguclu commented on chapter2/linear-elasticity-intro.md:
```diff
@@ -0,0 +1,157 @@
+
+with
+
+$$
+
+\begin{aligned}
+&a : 
+\begin{cases}
+V \times V \rightarrow \mathbb{R} \\
+(u, v) \longmapsto \int_{\Omega} \epsilon(u) : C : \epsilon(v) \, \mathrm{d} x
+\end{cases} \\[0.3cm]
+&L : 
+\begin{cases}
+V \rightarrow \mathbb{R} \\
+v \longmapsto \int_{\Omega} f \cdot v \, \mathrm{d} x + \int_{\partial \Omega_T} g_T \cdot v \, \mathrm{d} s
+\end{cases}
+\end{aligned}
+
+$$
```
```suggestion
$$
\begin{aligned}
&a : 
\begin{cases}
V \times V \rightarrow \mathbb{R} \\
(u, v) \longmapsto \int_{\Omega} \epsilon(u) : C : \epsilon(v) \, \mathrm{d} x
\end{cases} \\[0.3cm]
&L : 
\begin{cases}
V \rightarrow \mathbb{R} \\
v \longmapsto \int_{\Omega} f \cdot v \, \mathrm{d} x + \int_{\partial \Omega_T} g_T \cdot v \, \mathrm{d} s
\end{cases}
\end{aligned}
$$
```

yguclu commented on chapter2/linear-elasticity-intro.md:
```diff
@@ -0,0 +1,157 @@
+## Isotropic Materials
+For isotropic materials, the elasticity tensor $C$ can be expressed in terms of the Lamé parameters $\lambda$ and $\mu$ as follows:
+
+$$
+
+C := \lambda (\nabla \cdot u) I_3 + 2\mu \epsilon(u)
+
+$$
```
```suggestion
$$
C := \lambda (\nabla \cdot u) I_3 + 2\mu \epsilon(u)
$$
```

yguclu commented on chapter2/linear-elasticity-intro.md:
```diff
@@ -0,0 +1,157 @@
+
+
+This leads to the variational formulation:
+$$
+
+\boxed{
+\begin{aligned}
+&\text{Find } u \in V \text{ such that:}
+\\
+&\qquad a(u, v) = L(v) \quad \forall v \in V
+\end{aligned}
+}
+
+$$
```
```suggestion
$$
\boxed{
\begin{aligned}
&\text{Find } u \in V \text{ such that:}
\\
&\qquad a(u, v) = L(v) \quad \forall v \in V
\end{aligned}
}
$$
```

yguclu commented on chapter2/linear-elasticity-intro.md:
```diff
@@ -0,0 +1,157 @@
+$$
+
+with
+$$
+
+\begin{aligned}
+&a :\begin{cases}
+V \times V \rightarrow \mathbb{R} \\
+(u, v) \longmapsto \int_{\Omega} \sigma(u) : \epsilon(v) \, \mathrm{d} x
+\end{cases} \\[0.3cm]
+&L :\begin{cases}
+V \rightarrow \mathbb{R} \\
+v \longmapsto \int_{\Omega} f \cdot v \, \mathrm{d} x + \int_{\partial \Omega_T} g_T \cdot v \, \mathrm{d} s
+\end{cases}
+\end{aligned}
+
+$$
```
```suggestion
$$
\begin{aligned}
&a :\begin{cases}
V \times V \rightarrow \mathbb{R} \\
(u, v) \longmapsto \int_{\Omega} \sigma(u) : \epsilon(v) \, \mathrm{d} x
\end{cases} \\[0.3cm]
&L :\begin{cases}
V \rightarrow \mathbb{R} \\
v \longmapsto \int_{\Omega} f \cdot v \, \mathrm{d} x + \int_{\partial \Omega_T} g_T \cdot v \, \mathrm{d} s
\end{cases}
\end{aligned}
$$
```

yguclu reviewed (changes requested):
Your math equations do not get formatted on the GitHub website. This makes it very hard for me to review the PR. I suggested some changes on the intro file which hopefully fix the problem.

yguclu committed:
Apply suggestions from code review

Let the equations be formatted on GitHub...

yguclu commented on chapter2/linear-elasticity-intro.md:
```diff
@@ -0,0 +1,137 @@
+- $\partial \Omega_D$ is the Dirichlet boundary condition where the displacement is fixed to zero, and $\partial \Omega_T$ is the Neumann boundary condition where the traction is prescribed.
+
+## The Variational Formulation
+The variational formulation of the linear elasticity equations involves forming the inner product of the PDE with a vector test function $ v \in V $ and integrating over the domain $ \Omega $. This yields:
```
Empty spaces next to the `$` delimiters stop GitHub from formatting inline math.
```suggestion
The variational formulation of the linear elasticity equations involves forming the inner product of the PDE with a vector test function $v \in V$ and integrating over the domain $\Omega$. This yields:
```

yguclu commented on chapter2/linear-elasticity-intro.md:
```diff
@@ -0,0 +1,137 @@
+\int_{\Omega} - \nabla \cdot \sigma(u) \cdot v \, \mathrm{d} x = \int_{\Omega} f \cdot v \, \mathrm{d} x
+$$
+
+Integrating the term $ \nabla \cdot \sigma(u) \cdot v $ by parts, considering boundary conditions, we obtain:
```
See previous comment.
```suggestion
Integrating the term $\nabla \cdot \sigma(u) \cdot v$ by parts, considering boundary conditions, we obtain:
```

yguclu commented on chapter2/linear-elasticity-intro.md:
```diff
@@ -0,0 +1,137 @@
+\int_{\Omega} \sigma(u) : \nabla v \, \mathrm{d} x = \int_{\Omega} f \cdot v \, \mathrm{d} x + \int_{\partial \Omega_T} g_T \cdot v \, \mathrm{d} s
+$$
+
+By using the symmetry of the stress tensor $ \sigma $ and its definition from $(2)$, we can notice that :
```
```suggestion
By using the symmetry of the stress tensor $\sigma$ and its definition from $(2)$, we can notice that :
```

yguclu commented on chapter2/linear-elasticity-intro.md:
```diff
@@ -0,0 +1,137 @@
+$$\sigma(u) = \lambda (\nabla \cdot u) I_3 + 2\mu \epsilon(u)$$
+
+
+This leads to the variational formulation:
+$$
```
We need an empty line between text and the math environment.
```suggestion
This leads to the variational formulation:

$$
```

yguclu commented on chapter2/linear-elasticity-intro.md:
```diff
@@ -0,0 +1,137 @@
+}
+$$
+
+with
+$$
```
We need an empty line between text and the math environment.
```suggestion
with

$$
```

yguclu reviewed (changes requested):
There are still some formatting problems...

yguclu committed:
Apply suggestions from code review

Fix some more formatting issues

yguclu reviewed (changes requested):
@ArasuCandassamy Good job with this PR!

I have just made some changes to the intro markdown file in order to be able to properly visualize the inline math and the equations on GitHub (when I click on "view file"). Could you please make similar changes to the other two markdown files?

ArasuCandassamy committed:
Solving formatting issues

ArasuCandassamy committed:
Solving formatting issues

ArasuCandassamy committed:
Cleaning notebooks
